pkg/edgeview/src/basics.go

```diff
@@ -0,0 +1,665 @@
+				log.Errorf("received invalid IP %v", msg)
+				return false
+			}
+			if len(addrPort) == 2 {
+				num, err := strconv.Atoi(addrPort[1])
+				if err != nil || num < 1024 || num > 65535 {
+					log.Errorf("received invalid IP %v", msg)
+					return false
+				}
+			} else if len(addrPort) > 2 {
+				log.Errorf("received invalid IP %v", msg)
+				return false
+			}
+			myEvEndPoint = msgs[1]
+			return true
+		}
+	}
+	return false
+}
+
+func evLogger(deb bool) *logrus.Logger {
+	formatter := logrus.JSONFormatter{
+		TimestampFormat: time.RFC3339Nano,
+	}
+	logger := logrus.New()
+	logger.SetFormatter(&formatter)
+	log = base.NewSourceLogObject(logger, agentName, os.Getpid())
+	if deb {
+		logger.SetLevel(logrus.DebugLevel)
+	}
+	return logger
+}
+
+func getBasics() {
+	if runOnServer {
+		if _, err := os.Stat("/config"); err != nil {
+			return
+		}
+	}
+
+	ips := []string{}
+	for _, i := range allUPIntfIPv4() {
+		if strings.HasPrefix(i.intfName, "bn") {
+			continue
+		}
+		ips = append(ips, i.ipAddr)
+	}
+	if myEvEndPoint != "" {
+		fmt.Printf("Device IPs: %v; Endpoint IP %s\n", ips, myEvEndPoint)
+	} else {
+		fmt.Printf("Device IPs: %v\n", ips)
+	}
+
+	retbytes, err := ioutil.ReadFile("/persist/status/uuid")
+	if err == nil {
+		fmt.Printf("  UUID: %s", string(retbytes))
+	}
+	retbytes, err = ioutil.ReadFile("/config/server")
+	if err == nil {
+		var printed bool
+		serverurl := string(retbytes)
+		conts := strings.Split(serverurl, "zedcloud.")
+		if len(conts) == 2 {
+			cont2s := strings.Split(conts[1], ".zededa.net")
+			if len(cont2s) == 2 {
+				cluster := cont2s[0]
+				controller := strings.Replace(serverurl, cluster, "\033[0;93m"+cluster+"\033[0m", 1)
+				fmt.Printf("  Controller: %s", controller)
+				printed = true
+			}
+		}
+		if !printed {
+			fmt.Printf("  Controller: %s", serverurl)
+		}
+	}
+
+	var release, partition string
+	retbytes, err = ioutil.ReadFile("/run/eve-release")
+	if err == nil {
+		release = string(retbytes)
+	}
+	retbytes, err = ioutil.ReadFile("/run/eve.id")
+	if err == nil {
+		partition = string(retbytes)
+	}
+
+	fmt.Printf("  EVE-OS release %s, %s", release, partition)
+	fmt.Printf("  Edge-View Ver: %s, JWT expires at %v\n",
+		edgeViewVersion, time.Unix(int64(evStatus.ExpireOn), 0).Format(time.RFC3339))
+	hinfo, err := host.Info()
+	if err == nil {
+		loc, _ := time.LoadLocation("UTC")
+		fmt.Printf("  %v(%v), uptime %d (sec) = %d days",
+			time.Now().In(loc).Format(time.RFC3339), loc, hinfo.Uptime, hinfo.Uptime/(3600*24))
+	}
+	fmt.Println()
+	fmt.Println()
+	closePipe(true)
+}
+
+// dynamically install package for uncommon options
+func addPackage(pathName, pkgName string) error {
+	_, err := os.Stat(pathName)
+	if err != nil {
+		if os.IsNotExist(err) {
+			_, err = runCmd("apk add "+pkgName, false)
+			if err != nil {
+				fmt.Printf("%v\n", err)
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+func runCmd(cmd string, isPrint bool) (string, error) {
```

Comment:
Currenly the caller need to create a string with whitespace and it is hard to read to see that there is whitespace between the different arguments. Why not make this funftion (and runCmdInFunction) take the same approach as exec.Command i.e., prog string, args []string. That reduces the risk of whitespace errors and the need for a Fields call inside runCmdInFunction.

It could be that some callers which variable number of args need to call string.Fields but then at least it is explicit which ones pass in more complex lists of arguments and which ones have a fixed number.

Reply:
changed to runCmdInFunction()

Reply:
I don't see a change which relates to by comment to use []string instead of string.

Reply:
ok, i misunderstood. let me change.